tests/integrations/raftstore/test_unsafe_recovery.rs

```diff
@@ -0,0 +1,544 @@
+    let read_index = new_read_index_cmd();
+    must_get_error_recovery_in_progress(&mut cluster, &region, read_index);
+
+    cluster.exit_force_leader(region.get_id(), 1);
+
+    // quorum is formed, can propose command successfully now
+    cluster.must_put(b"k4", b"v4");
+    assert_eq!(cluster.must_get(b"k2"), None);
+    assert_eq!(cluster.must_get(b"k3"), None);
+    assert_eq!(cluster.must_get(b"k4"), Some(b"v4".to_vec()));
+}
+
+// Test the case that three of five nodes fail and force leader on the rest node
+// which is a learner.
+#[test]
+fn test_force_leader_for_learner() {
+    let mut cluster = new_node_cluster(0, 5);
+    cluster.pd_client.disable_default_operator();
+
+    cluster.run();
+    cluster.must_put(b"k1", b"v1");
+
+    let region = cluster.get_region(b"k1");
+    cluster.must_split(&region, b"k9");
+    let mut region = cluster.get_region(b"k2");
+    let peer_on_store5 = find_peer(&region, 5).unwrap();
+    cluster.must_transfer_leader(region.get_id(), peer_on_store5.clone());
+
+    let peer_on_store1 = find_peer(&region, 1).unwrap();
+    // replace one peer with learner
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), peer_on_store1.clone());
+    cluster.pd_client.must_add_peer(
+        region.get_id(),
+        new_learner_peer(peer_on_store1.get_store_id(), peer_on_store1.get_id()),
+    );
+
+    must_get_equal(&cluster.get_engine(1), b"k1", b"v1");
+
+    cluster.stop_node(3);
+    cluster.stop_node(4);
+    cluster.stop_node(5);
+
+    let put = new_put_cmd(b"k2", b"v2");
+    let req = new_request(region.get_id(), region.take_region_epoch(), vec![put], true);
+    // quorum is lost, can't propose command successfully.
+    assert!(
+        cluster
+            .call_command_on_leader(req, Duration::from_millis(10))
+            .is_err()
+    );
+
+    cluster.enter_force_leader(region.get_id(), 1, HashSet::from_iter(vec![3, 4, 5]));
+    // promote the learner first and remove the peers on failed nodes
+    cluster
+        .pd_client
+        .must_add_peer(region.get_id(), find_peer(&region, 1).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 3).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 4).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 5).unwrap().clone());
+    cluster.exit_force_leader(region.get_id(), 1);
+
+    // quorum is formed, can propose command successfully now
+    cluster.must_put(b"k4", b"v4");
+    assert_eq!(cluster.must_get(b"k2"), None);
+    assert_eq!(cluster.must_get(b"k3"), None);
+    assert_eq!(cluster.must_get(b"k4"), Some(b"v4".to_vec()));
+    cluster.must_transfer_leader(region.get_id(), find_peer(&region, 1).unwrap().clone());
+}
+
+// Test the case that none of five nodes fails and force leader on one of the nodes.
```

Comment:
The case may provide the wrong impression that force_leader will not break correctness if triggered accidentally. Should comment its limitation.

Reply:
It doesn't break the correctness. Pre-force-leader check probits it to become force leader.

Reply:
How? For example, a group of a, b and c. c is isolated from a, a is the leader. If c has increased its term by 2 somehow (for example false prevote success two time) and force leader is sent to b, then b will reject a's heartbeat while can vote for c. So c becomes leader and there are two leaders in the group.

Reply:
but b won't become leader because it receives vote from an unexpected peer.